File: core/provenance/provenance.py
#!/usr/bin/env python3
"""
provenance.py — the shared spine of Sentinel OS.

Every tool in this repo already invented the same discipline independently:
  * PRA               — metadata-only records, SHA-256 manifests, append-only ledgers
  * OpenMontage       — "every render emits a manifest: input hashes, output hash"
  * Atlas Vuln        — "nothing enters the record without a source and a hash"

This module is that discipline written once, so a received record, a rendered
video, and a CVE report all describe their provenance in ONE shape. That shared
shape is what lets the PRA evidence database index every artifact the whole
system produces, not just public-records files.

A provenance record answers four questions about an artifact:
  WHAT is it        — kind, id, a human label
  WHERE did it come from — source url / local path / upstream artifact ids
  IS IT INTACT      — sha256 of the exact bytes
  WHEN / BY WHAT    — utc timestamp, tool + version, and the inputs that made it

Design rules (match the rest of the repo):
  * Pure standard library. No dependencies.
  * Paths recorded RELATIVE to an evidence root — never absolute. An absolute
    path leaks the operator's machine layout and breaks portability.
  * The ledger is append-only. This module never rewrites or deletes a line.
  * Nothing here reaches the network.
"""
from __future__ import annotations

import hashlib
import json
import os
import re
from datetime import datetime, timezone
from typing import Any, Dict, Iterable, List, Optional

SCHEMA_VERSION = "provenance/1"
_ABS_PATH = re.compile(r"^([A-Za-z]:[\\/]|/|\\\\)")


# ---------------------------------------------------------------------------
# Hashing
# ---------------------------------------------------------------------------

def sha256_bytes(data: bytes) -> str:
    """SHA-256 of a byte string, lowercase hex."""
    return hashlib.sha256(data).hexdigest()


def sha256_file(path: str, _chunk: int = 1024 * 1024) -> str:
    """SHA-256 of a file, streamed so large videos don't blow up memory."""
    h = hashlib.sha256()
    with open(path, "rb") as fh:
        for block in iter(lambda: fh.read(_chunk), b""):
            h.update(block)
    return h.hexdigest()


def sha256_text(text: str) -> str:
    """SHA-256 of text, UTF-8 encoded."""
    return sha256_bytes(text.encode("utf-8"))


def sha256_json(obj: Any) -> str:
    """
    Stable SHA-256 of a JSON-serializable object. Keys are sorted and separators
    fixed, so the same logical object always hashes identically regardless of
    dict insertion order.
    """
    return sha256_text(json.dumps(obj, sort_keys=True, separators=(",", ":")))


# ---------------------------------------------------------------------------
# Paths
# ---------------------------------------------------------------------------

def is_absolute(path: str) -> bool:
    return bool(_ABS_PATH.match(path or ""))


def relativize(path: str, evidence_root: Optional[str]) -> str:
    """
    Express `path` relative to the evidence root. Refuses to emit an absolute
    path — an artifact record must be portable and must not leak machine layout.
    """
    if not path:
        return path
    if evidence_root:
        try:
            rel = os.path.relpath(os.path.abspath(path), os.path.abspath(evidence_root))
            if not rel.startswith(".."):
                return rel
        except ValueError:
            pass  # different drive on Windows, fall through
    if is_absolute(path):
        # Keep only the basename rather than record an absolute path.
        return os.path.basename(path)
    return path


# ---------------------------------------------------------------------------
# Sourcing tiers — carried over from the Flock production GlassMark scheme,
# generalized so every module speaks the same sourcing language.
# ---------------------------------------------------------------------------

TIERS = {
    "GREEN":        "primary document/artifact in custody (we hold the file and its hash)",
    "ATTRIBUTED":   "another party's material, credited to its source",
    "SOURCE_NEEDED": "not verified yet — must not be published or asserted",
    "GENERATED":    "produced by this system (a render, a report), traceable to its inputs",
    "NA":           "no factual claim attached",
}


def valid_tier(tier: str) -> bool:
    return tier in TIERS


# ---------------------------------------------------------------------------
# The provenance record
# ---------------------------------------------------------------------------

def utc_now() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


def make_record(
    *,
    kind: str,
    artifact_id: str,
    label: str = "",
    tool: str = "",
    tool_version: str = "",
    tier: str = "GENERATED",
    sha256: Optional[str] = None,
    local_path: Optional[str] = None,
    evidence_root: Optional[str] = None,
    source_url: Optional[str] = None,
    source_ref: Optional[str] = None,
    inputs: Optional[Iterable[Dict[str, Any]]] = None,
    extra: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """
    Build one provenance record.

    kind         'received_record' | 'video_build' | 'vuln_report' | 'source' | ...
    artifact_id  stable id within its kind (a request id, a build id, a report id)
    tier         a sourcing tier (see TIERS); defaults to GENERATED
    sha256       hash of the artifact bytes, if it is a file/blob
    local_path   where the artifact lives, recorded RELATIVE to evidence_root
    inputs       list of {artifact_id?, sha256?, path?, note?} that produced it
    extra        module-specific fields (kept in their own namespace)

    Raises on an invalid tier — a wrong tier is worse than no tier.
    """
    if not kind or not artifact_id:
        raise ValueError("provenance record requires kind and artifact_id")
    if not valid_tier(tier):
        raise ValueError(f"unknown sourcing tier {tier!r}; valid: {sorted(TIERS)}")

    rec: Dict[str, Any] = {
        "schema": SCHEMA_VERSION,
        "kind": kind,
        "artifact_id": artifact_id,
        "label": label or "",
        "tier": tier,
        "recorded_at": utc_now(),
        "tool": tool or "",
        "tool_version": tool_version or "",
    }
    if sha256:
        rec["sha256"] = sha256
    if local_path is not None:
        rel = relativize(local_path, evidence_root)
        if is_absolute(rel):
            raise ValueError(f"refusing to record an absolute path: {local_path!r}")
        rec["local_path"] = rel
    if source_url:
        rec["source_url"] = source_url
    if source_ref:
        rec["source_ref"] = source_ref

    norm_inputs: List[Dict[str, Any]] = []
    for i in (inputs or []):
        item = {k: v for k, v in i.items() if v is not None}
        if "path" in item:
            item["path"] = relativize(item["path"], evidence_root)
        norm_inputs.append(item)
    if norm_inputs:
        rec["inputs"] = norm_inputs

    if extra:
        rec["extra"] = extra

    # A content hash of the record itself (minus its own hash field), so the
    # ledger line is tamper-evident the same way every module already is.
    rec["record_sha256"] = sha256_json({k: v for k, v in rec.items()})
    return rec


# ---------------------------------------------------------------------------
# The append-only ledger
# ---------------------------------------------------------------------------

class Ledger:
    """
    Append-only JSONL provenance ledger. One record per line. This class never
    rewrites or truncates the file — same guarantee PRA enforces at the database
    level with its append-only triggers.
    """

    def __init__(self, path: str):
        self.path = path
        os.makedirs(os.path.dirname(os.path.abspath(path)), exist_ok=True)

    def append(self, record: Dict[str, Any]) -> Dict[str, Any]:
        if "record_sha256" not in record:
            raise ValueError("record must be built with make_record()")
        with open(self.path, "a", encoding="utf-8") as fh:
            fh.write(json.dumps(record, sort_keys=True, separators=(",", ":")) + "\n")
        return record

    def read_all(self) -> List[Dict[str, Any]]:
        if not os.path.exists(self.path):
            return []
        out = []
        with open(self.path, "r", encoding="utf-8") as fh:
            for line in fh:
                line = line.strip()
                if line:
                    out.append(json.loads(line))
        return out

    def verify(self) -> Dict[str, Any]:
        """
        Recompute every line's record_sha256 and report any that don't match —
        i.e. lines that were edited in place after they were written.
        """
        rows = self.read_all()
        tampered = []
        for idx, row in enumerate(rows):
            claimed = row.get("record_sha256")
            recomputed = sha256_json({k: v for k, v in row.items() if k != "record_sha256"})
            if claimed != recomputed:
                tampered.append({"line": idx + 1, "artifact_id": row.get("artifact_id")})
        return {"total": len(rows), "tampered": tampered, "ok": not tampered}


if __name__ == "__main__":
    # Tiny smoke test so `python3 provenance.py` proves the module runs.
    r = make_record(kind="video_build", artifact_id="demo-001", label="smoke test",
                    tool="provenance.py", tool_version=SCHEMA_VERSION,
                    tier="GENERATED", sha256=sha256_text("hello"))
    print(json.dumps(r, indent=2))

File: core/provenance/provenance.js
'use strict';
/**
 * provenance.js — the JavaScript half of the Sentinel OS provenance spine.
 *
 * This is a deliberate twin of provenance.py, not a loose reimplementation.
 * The Node modules (PRA persistence, connectors, ingest) and the Python modules
 * (OpenMontage, atlas-vuln, analysis) both write into the same evidence store,
 * so a record built on either side must hash to the same value. If these two
 * files ever disagree, the two halves of the system can no longer verify each
 * other's chain of custody — so test_provenance.py builds the same record in
 * both languages and compares the hashes.
 *
 * Same rules as the Python side: standard library only, relative paths only,
 * append-only ledger, no network.
 */

const crypto = require('crypto');
const fs = require('fs');
const path = require('path');

const SCHEMA_VERSION = 'provenance/1';
const ABS_PATH = /^([A-Za-z]:[\\/]|\/|\\\\)/;

// ---------------------------------------------------------------------------
// Canonical JSON
//
// Must byte-match Python's json.dumps(obj, sort_keys=True, separators=(',',':')).
// Two details that JSON.stringify gets wrong for our purposes:
//   1. it does not sort keys, and Python sorts them at every nesting level
//   2. Python defaults to ensure_ascii=True, escaping every non-ASCII character
// Both are reproduced here.
// ---------------------------------------------------------------------------

function escapeString(s) {
  let out = '"';
  for (const ch of s) {
    const code = ch.codePointAt(0);
    if (ch === '"') out += '\\"';
    else if (ch === '\\') out += '\\\\';
    else if (ch === '\n') out += '\\n';
    else if (ch === '\r') out += '\\r';
    else if (ch === '\t') out += '\\t';
    else if (ch === '\b') out += '\\b';
    else if (ch === '\f') out += '\\f';
    else if (code < 0x20) out += '\\u' + code.toString(16).padStart(4, '0');
    else if (code < 0x7f) out += ch;
    else if (code <= 0xffff) out += '\\u' + code.toString(16).padStart(4, '0');
    else {
      // Python emits a surrogate pair for astral characters.
      const v = code - 0x10000;
      const hi = 0xd800 + (v >> 10);
      const lo = 0xdc00 + (v & 0x3ff);
      out += '\\u' + hi.toString(16).padStart(4, '0') +
             '\\u' + lo.toString(16).padStart(4, '0');
    }
  }
  return out + '"';
}

function canonicalJson(value) {
  if (value === null) return 'null';
  const t = typeof value;
  if (t === 'boolean') return value ? 'true' : 'false';
  if (t === 'number') {
    if (!Number.isFinite(value)) throw new TypeError('non-finite number is not canonical JSON');
    return Number.isInteger(value) ? String(value) : String(value);
  }
  if (t === 'string') return escapeString(value);
  if (Array.isArray(value)) return '[' + value.map(canonicalJson).join(',') + ']';
  if (t === 'object') {
    const keys = Object.keys(value).filter((k) => value[k] !== undefined).sort();
    return '{' + keys.map((k) => escapeString(k) + ':' + canonicalJson(value[k])).join(',') + '}';
  }
  throw new TypeError('cannot canonicalize ' + t);
}

// ---------------------------------------------------------------------------
// Hashing
// ---------------------------------------------------------------------------

function sha256Bytes(buf) {
  return crypto.createHash('sha256').update(buf).digest('hex');
}

function sha256Text(text) {
  return sha256Bytes(Buffer.from(text, 'utf8'));
}

function sha256Json(obj) {
  return sha256Text(canonicalJson(obj));
}

/** SHA-256 of a file, streamed so large videos don't blow up memory. */
function sha256File(filePath) {
  const h = crypto.createHash('sha256');
  const fd = fs.openSync(filePath, 'r');
  try {
    const buf = Buffer.alloc(1024 * 1024);
    let n;
    while ((n = fs.readSync(fd, buf, 0, buf.length, null)) > 0) h.update(buf.subarray(0, n));
  } finally {
    fs.closeSync(fd);
  }
  return h.digest('hex');
}

// ---------------------------------------------------------------------------
// Paths
// ---------------------------------------------------------------------------

function isAbsolute(p) {
  return ABS_PATH.test(p || '');
}

/**
 * Express `p` relative to the evidence root. Refuses to emit an absolute path —
 * an artifact record must be portable and must not leak machine layout.
 */
function relativize(p, evidenceRoot) {
  if (!p) return p;
  if (evidenceRoot) {
    const rel = path.relative(path.resolve(evidenceRoot), path.resolve(p));
    if (rel && !rel.startsWith('..')) return rel;
  }
  if (isAbsolute(p)) return path.basename(p);
  return p;
}

// ---------------------------------------------------------------------------
// Sourcing tiers — must stay identical to provenance.py TIERS
// ---------------------------------------------------------------------------

const TIERS = {
  GREEN: 'primary document/artifact in custody (we hold the file and its hash)',
  ATTRIBUTED: "another party's material, credited to its source",
  SOURCE_NEEDED: 'not verified yet — must not be published or asserted',
  GENERATED: 'produced by this system (a render, a report), traceable to its inputs',
  NA: 'no factual claim attached',
};

function validTier(tier) {
  return Object.prototype.hasOwnProperty.call(TIERS, tier);
}

// ---------------------------------------------------------------------------
// The provenance record
// ---------------------------------------------------------------------------

function utcNow() {
  return new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
}

/**
 * Build one provenance record. Mirrors make_record() in provenance.py field for
 * field; the camelCase argument names are the only intentional difference.
 *
 * Throws on an invalid tier — a wrong tier is worse than no tier.
 */
function makeRecord(opts) {
  const {
    kind, artifactId, label = '', tool = '', toolVersion = '',
    tier = 'GENERATED', sha256 = null, localPath = null, evidenceRoot = null,
    sourceUrl = null, sourceRef = null, inputs = null, extra = null,
  } = opts || {};

  if (!kind || !artifactId) throw new Error('provenance record requires kind and artifactId');
  if (!validTier(tier)) {
    throw new Error(`unknown sourcing tier ${JSON.stringify(tier)}; valid: ${Object.keys(TIERS).sort().join(', ')}`);
  }

  const rec = {
    schema: SCHEMA_VERSION,
    kind,
    artifact_id: artifactId,
    label: label || '',
    tier,
    recorded_at: utcNow(),
    tool: tool || '',
    tool_version: toolVersion || '',
  };
  if (sha256) rec.sha256 = sha256;
  if (localPath !== null && localPath !== undefined) {
    const rel = relativize(localPath, evidenceRoot);
    if (isAbsolute(rel)) throw new Error(`refusing to record an absolute path: ${localPath}`);
    rec.local_path = rel;
  }
  if (sourceUrl) rec.source_url = sourceUrl;
  if (sourceRef) rec.source_ref = sourceRef;

  const normInputs = [];
  for (const i of inputs || []) {
    const item = {};
    for (const [k, v] of Object.entries(i)) {
      if (v !== null && v !== undefined) item[k] = v;
    }
    if ('path' in item) item.path = relativize(item.path, evidenceRoot);
    normInputs.push(item);
  }
  if (normInputs.length) rec.inputs = normInputs;

  if (extra) rec.extra = extra;

  rec.record_sha256 = sha256Json({ ...rec });
  return rec;
}

// ---------------------------------------------------------------------------
// The append-only ledger
// ---------------------------------------------------------------------------

class Ledger {
  constructor(ledgerPath) {
    this.path = ledgerPath;
    fs.mkdirSync(path.dirname(path.resolve(ledgerPath)), { recursive: true });
  }

  append(record) {
    if (!record || !record.record_sha256) throw new Error('record must be built with makeRecord()');
    fs.appendFileSync(this.path, canonicalJson(record) + '\n', 'utf8');
    return record;
  }

  readAll() {
    if (!fs.existsSync(this.path)) return [];
    return fs.readFileSync(this.path, 'utf8')
      .split('\n')
      .map((l) => l.trim())
      .filter(Boolean)
      .map((l) => JSON.parse(l));
  }

  /** Recompute every line's record_sha256 and report lines edited after writing. */
  verify() {
    const rows = this.readAll();
    const tampered = [];
    rows.forEach((row, idx) => {
      const claimed = row.record_sha256;
      const body = { ...row };
      delete body.record_sha256;
      if (claimed !== sha256Json(body)) {
        tampered.push({ line: idx + 1, artifact_id: row.artifact_id });
      }
    });
    return { total: rows.length, tampered, ok: tampered.length === 0 };
  }
}

module.exports = {
  SCHEMA_VERSION, TIERS,
  canonicalJson, sha256Bytes, sha256Text, sha256Json, sha256File,
  isAbsolute, relativize, validTier, utcNow, makeRecord, Ledger,
};
